dragonfly/model.py

```diff
@@ -0,0 +1,357 @@
+# coding: utf-8
+"""Dragonfly Model."""
+from ._base import _BaseGeometry
+from .properties import ModelProperties
+from .building import Building
+from .context import ContextShade
+
+from honeybee.shade import Shade
+from honeybee.boundarycondition import Surface
+from honeybee.facetype import AirWall
+from honeybee.typing import float_in_range
+
+from ladybug_geometry.geometry2d.pointvector import Vector2D
+
+import math
+
+
+class Model(_BaseGeometry):
+    """A collection of Buildings and ContextShades for an entire model.
+
+    Properties:
+        * name
+        * display_name
+        * north_angle
+        * north_vector
+        * buildings
+        * context_shades
+    """
+    __slots__ = ('_buildings', '_context_shades', '_north_angle', '_north_vector')
+
+    def __init__(self, name, buildings=None, context_shades=None, north_angle=0):
+        """A collection of Buildings and ContextShades for an entire model.
+
+        Args:
+            name: Model name. Must be < 100 characters.
+            buildings: A list of Building objects in the model.
+            context_shades: A list of ContextShade objects in the model.
+            north_angle: An number between 0 and 360 to set the clockwise north
+                direction in degrees. Default is 0.
+        """
+        self.name = name
+        self.north_angle = north_angle
+
+        self._buildings = []
+        self._context_shades = []
+        if buildings is not None:
+            for bldg in buildings:
+                self.add_building(bldg)
+        if context_shades is not None:
+            for shade in context_shades:
+                self.add_context_shade(shade)
+
+        self._properties = ModelProperties(self)
+
+    @classmethod
+    def from_dict(cls, data):
+        """Initialize a Model from a dictionary.
+
+        Args:
+            data: A dictionary representation of a Model object.
+        """
+        # check the type of dictionary
+        assert data['type'] == 'Model', 'Expected Model dictionary. ' \
+            'Got {}.'.format(data['type'])
+
+        buildings = None  # import buildings
+        if 'buildings' in data and data['buildings'] is not None:
+            buildings = [Building.from_dict(bldg) for bldg in data['buildings']]
+        context_shades = None  # import context shades
+        if 'context_shades' in data and data['context_shades'] is not None:
+            context_shades = [ContextShade.from_dict(s) for s in data['context_shades']]
+
+        # import the north angle
+        north_angle = 0 if 'north_angle' not in data else data['north_angle']
+
+        # build the model object
+        model = Model(data['name'], buildings, context_shades, north_angle)
+        assert model.display_name == model.name, \
```

Comment:
I wish there was a way to do this check first and give an error before converting all the buildings, etc.

Reply:
There is but it requires a couple of extra lines of code and an extra processing of the name string. The way it is here just seemed more elegant to me, particularly because I expect this to be the exception and not the rule. But I can do this check up front if you think that it will be pretty common.